nomad/structs/structs.go

```diff
@@ -950,6 +950,10 @@ type ApplyPlanResultsRequest struct {
 // to cause evictions or to assign new allocations. Both can be done
 // within a single transaction
 type AllocUpdateRequest struct {
+	// NodeID is the ID of the node sending the request.
+	// It may be empty if this is not an RPC request from a client.
```

Comment:
Hm I'd really like to avoid more conditionally-initialized fields as this is exactly what makes figuring out things like `stripped allocs sent from clients` so hard.

AFAICT this struct is used for *three mostly disjoint requests:*

1. Stripped allocs from clients Node.UpdateAlloc RPC using the Allocs, NodeID (new), and WriteRequest fields
2. Raft log message using the Allocs, Evals, and WriteRequest fields
3. Plan updates using the AllocsStopped, AllocsUpdated, and Job fields

1 & 2 using the same alloc is defensible but 3 just makes no sense these days. 3 isn't your problem though at all.

I can't think of a way to change this that I'd be comfortable backporting, so maybe let's put a big ugly comment here and file an issue to clean this up someday? I really don't want to hold up your PR with refactoring this mess.

Reply:
Removed the new field as per suggestion [above](https://github.com/hashicorp/nomad/pull/15068#discussion_r1013503431). Thanks!